# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmSerialize.cpp
# Commit: bd2b8b0c2f35
# Full Hash: bd2b8b0c2f35ff48f08561f9a4d88f1f1ee12218
# Author: Julien Pages <jpages@mozilla.com>
# Date: 2025-03-04 04:58:00
# Regressor Bug: 1861078
# File Overlap Count: 1
# Description:
#   Bug 1861078 - wasm: Refactor stackmaps to use a hashmap instead of a vector. r=rhunt
#   
#   Wasm stackmaps were stored in a sorted vector and traced at runtime with multiple
#   binary searches. Using hashmaps should reduce the complexity of iterating over stackmaps.
#   
# ==============================================================================

diff -r 1a3bc285395f -r bd2b8b0c2f35 js/src/wasm/WasmSerialize.cpp
--- a/js/src/wasm/WasmSerialize.cpp	Mon Mar 03 16:37:47 2025 +0000
+++ b/js/src/wasm/WasmSerialize.cpp	Mon Mar 03 16:37:48 2025 +0000
@@ -1042,25 +1042,15 @@
   return Ok();
 }
 
-static inline uint32_t ComputeCodeOffset(const uint8_t* codeStart,
-                                         const uint8_t* codePtr) {
-  MOZ_RELEASE_ASSERT(codePtr >= codeStart);
-#ifdef JS_64BIT
-  MOZ_RELEASE_ASSERT(codePtr < codeStart + UINT32_MAX);
-#endif
-  return (uint32_t)(codePtr - codeStart);
-}
-
 CoderResult CodeStackMaps(Coder<MODE_DECODE>& coder,
-                          CoderArg<MODE_DECODE, wasm::StackMaps> item,
-                          const uint8_t* codeStart) {
-  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::StackMaps, 48);
+                          CoderArg<MODE_DECODE, wasm::StackMaps> item) {
+  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::StackMaps, 40);
   // Decode the amount of stack maps
   size_t length;
   MOZ_TRY(CodePod(coder, &length));
 
   for (size_t i = 0; i < length; i++) {
-    // Decode the offset relative to codeStart
+    // Decode the offset
     uint32_t codeOffset;
     MOZ_TRY(CodePod(coder, &codeOffset));
 
@@ -1069,37 +1059,32 @@
     MOZ_TRY(CodeStackMap(coder, &map));
 
     // Add it to the map
-    const uint8_t* nextInsnAddr = codeStart + codeOffset;
-    if (!item->add(nextInsnAddr, map)) {
+    if (!item->add(codeOffset, map)) {
       return Err(OutOfMemory());
     }
   }
 
-  // Finish the maps, asserting they are sorted
-  item->finishAlreadySorted();
   return Ok();
 }
 
 template <CoderMode mode>
 CoderResult CodeStackMaps(Coder<mode>& coder,
-                          CoderArg<mode, wasm::StackMaps> item,
-                          const uint8_t* codeStart) {
-  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::StackMaps, 48);
+                          CoderArg<mode, wasm::StackMaps> item) {
+  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::StackMaps, 40);
   STATIC_ASSERT_ENCODING_OR_SIZING;
 
   // Encode the amount of stack maps
   size_t length = item->length();
   MOZ_TRY(CodePod(coder, &length));
 
-  for (size_t i = 0; i < length; i++) {
-    StackMaps::Maplet maplet = item->get(i);
-    uint32_t codeOffset = ComputeCodeOffset(codeStart, maplet.nextInsnAddr);
+  for (auto iter = item->mapping_.iter(); !iter.done(); iter.next()) {
+    uint32_t codeOffset = iter.get().key();
 
-    // Encode the offset relative to codeStart
+    // Encode the offset
     MOZ_TRY(CodePod(coder, &codeOffset));
 
     // Encode the stack map
-    MOZ_TRY(CodeStackMap(coder, maplet.map));
+    MOZ_TRY(CodeStackMap(coder, iter.get().value()));
   }
   return Ok();
 }
@@ -1352,7 +1337,7 @@
 CoderResult CodeCodeBlock(Coder<MODE_DECODE>& coder,
                           wasm::UniqueCodeBlock* item,
                           const wasm::LinkData& linkData) {
-  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::CodeBlock, 2584);
+  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::CodeBlock, 2576);
   *item = js::MakeUnique<CodeBlock>(CodeBlock::kindFromTier(Tier::Serialized));
   if (!*item) {
     return Err(OutOfMemory());
@@ -1368,7 +1353,7 @@
   MOZ_TRY(CodeCallSites(coder, &(*item)->callSites));
   MOZ_TRY(CodeTrapSites(coder, &(*item)->trapSites));
   MOZ_TRY(CodePodVector(coder, &(*item)->funcExports));
-  MOZ_TRY(CodeStackMaps(coder, &(*item)->stackMaps, (*item)->segment->base()));
+  MOZ_TRY(CodeStackMaps(coder, &(*item)->stackMaps));
   MOZ_TRY(CodePodVector(coder, &(*item)->tryNotes));
   MOZ_TRY(CodePodVector(coder, &(*item)->codeRangeUnwindInfos));
   return Ok();
@@ -1378,7 +1363,7 @@
 CoderResult CodeCodeBlock(Coder<mode>& coder,
                           CoderArg<mode, wasm::CodeBlock> item,
                           const wasm::LinkData& linkData) {
-  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::CodeBlock, 2584);
+  WASM_VERIFY_SERIALIZATION_FOR_SIZE(wasm::CodeBlock, 2576);
   STATIC_ASSERT_ENCODING_OR_SIZING;
   MOZ_TRY(Magic(coder, Marker::CodeBlock));
   // We don't support serializing sub-ranges yet. These only happen with
@@ -1391,7 +1376,7 @@
   MOZ_TRY(CodeCallSites(coder, &item->callSites));
   MOZ_TRY(CodeTrapSites(coder, &item->trapSites));
   MOZ_TRY(CodePodVector(coder, &item->funcExports));
-  MOZ_TRY(CodeStackMaps(coder, &item->stackMaps, item->segment->base()));
+  MOZ_TRY(CodeStackMaps(coder, &item->stackMaps));
   MOZ_TRY(CodePodVector(coder, &item->tryNotes));
   MOZ_TRY(CodePodVector(coder, &item->codeRangeUnwindInfos));
   return Ok();